### PR TITLE
Issue 188

### PR DESCRIPTION
**Pull request by magnus-madsen**
This PR adds a flag `--Xinvariants` to control whether `CheckClassAdapter` as discussed in #188 is enabled.

It also does a few minor refactorings, and re-enables TestExamples.

### CODE REVIEW

**Comment by magnus-madsen** [comment]
This PR is not complete yet. I will do a few refactorings before merging.

File: main/src/ca/uwaterloo/flix/language/phase/Codegen.scala [hunk 1]
```diff
@@ -159,19 +159,29 @@ object Codegen {
...
 
     // Initialize the visitor to create a class.
     visitor.visit(V1_8, ACC_PUBLIC + ACC_SUPER, decorate(ctx.prefix), null, asm.Type.getInternalName(Constants.objectClass), null)
 
     compileStaticFlixField(ctx, visitor)
     compileConstructor(ctx, visitor)
-    functions.foreach(compileFunction(ctx, visitor))
+    ctx.functions.foreach(compileFunction(ctx, visitor))
+
+    // Check class invariants, if compiler invariants are enabled.
```
Comment (mhyee) [inline main/src/ca/uwaterloo/flix/language/phase/Codegen.scala]: I'm pretty sure you always have to call `visitEnd()`.

Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/Codegen.scala]: Fixed.

File: main/test/ca/uwaterloo/flix/runtime/TestBackend.scala [hunk 2]
```diff
@@ -4189,14 +4185,12 @@ class TestBackend extends FunSuite {
...
-  // TODO: Typechecker doesn't properly handle ???
-  ignore("Expression.Binary - BinaryOperator.Implication.05") {
+  test("Expression.Binary - BinaryOperator.Implication.05") {
     val input = "def f: Bool = false ==> ???"
     val t = new Tester(input)
     t.runTest(Value.True, "f")
   }
 
-  // TODO: Typechecker doesn't properly handle ???
   ignore("Expression.Binary - BinaryOperator.Implication.06") {
```
Comment (mhyee) [inline main/test/ca/uwaterloo/flix/runtime/TestBackend.scala]: Missed un-ignoring this test, as well as the next one (UserError).

**Comment by magnus-madsen** [comment]
My plan is to merge this soon, and then I think we should aim to remove all ignored tests. Possibly by opening bug reports for each test we can't easily fix.

**Pull request merged**